src/ray/raylet/scheduling_resources.h

```diff
@@ -227,6 +279,14 @@ class ResourceIds {
   /// A vector of pairs of resource ID and a fraction of that ID (the fraction
   /// is at least zero and strictly less than 1).
   std::vector<std::pair<int64_t, double>> fractional_ids_;
+  /// A double to track the greatest id of the resource - this needs to be maintained to
+  /// ensure unique ids are used when creating new resources. This keeps incrementing,
+  /// and resources may have gaps in between. TODO(romilb): Is this a safe assumption?
```

Comment:
It'd be preferable not to have gaps. How can gaps arise?

Reply:
Gaps can arise when resources are deleted (through a call to `ResourceIds::DecreaseCapacity`). On deletion, we cannot simply decrement `greatest_id_` because the first available resource in the `whole_ids_` vector gets removed (which may not necessarily be the resource with the largest id).